Js module search depth

billti opened the pull request:
This is a possible solution for #6670 (load JavaScript files from `node_modules`).

A concern with searching node_modules and loading JavaScript code is just how much code we might suck in on a project with lots of dependencies (or even a few, if those have a large dependency chain). This adds a setting to specify how many "hops" to go when loading JavaScript files while searching `node_modules`.

As an example, just using Gulp, with a setting of `1` it loads all the JavaScript files in the 'gulp' package, but does not follow its top level imports. This gives intellisense such as the below.

<img width="345" alt="screen shot 2016-02-05 at 11 03 22 am" src="https://cloud.githubusercontent.com/assets/993909/12856231/b4e4d10c-cbf8-11e5-9cc7-afb533936c79.png">

Note how the gulp exports are visible, however `src` is showing as `any` as this is a re-export of another dependency. With a setting of `2` (or deeper), this resolves also, as shown below:

<img width="463" alt="screen shot 2016-02-05 at 11 04 10 am" src="https://cloud.githubusercontent.com/assets/993909/12856322/2acde3ea-cbf9-11e5-821a-ed621ca493a0.png">

A question is what to default this setting at. Even with just Gulp, this loads 147 JavaScript files if loading all dependencies. `0` seems unhelpful. Perhaps `1` is a reasonable default?

@vladima for feedback on the code, and @egamma for feedback on the defaults.

weswigham commented:
Out of curiosity, how feasible would it be to remove the limit but then (likely in the background) generate and cache a flattened definition file for each top level module (so we don't have to keep all that backing JS in the service)? Deps don't usually change all that frequently.

DanielRosenwasser commented on src/compiler/program.ts:
```diff
@@ -99,20 +106,27 @@ namespace ts {
+            }
+            if (typeof jsonContent.main === "string") {
+                const path = normalizePath(combinePaths(candidate, jsonContent.main));
+                const result = loadNodeModuleFromFile(/*don't add extension*/[""], path, failedLookupLocation, !directoryProbablyExists(getDirectoryPath(path), host), host);
```
Space after comment

mhegazy commented on src/compiler/program.ts:
```diff
@@ -65,7 +65,14 @@ namespace ts {
+            if(result && result.resolvedModule && result.resolvedModule.resolvedFileName){
+                if(!fileExtensionIs(result.resolvedModule.resolvedFileName, ".ts")){
+                    // Loaded a non-TypeScript file from a node modules search
+                    result.resolvedModule.isJsFileFromNodeSearch = true;
```
do not think you really need that. we can just keep the depth for all files.

billti replied:
Agreed. And after talking to Vlad and reviewing the code, I can likely also just use the flag he already has (`isExternalLibrary`).

mhegazy commented on src/compiler/program.ts:
```diff
@@ -99,20 +106,27 @@ namespace ts {
+                    return result;
+                }
+            }
+            if (typeof jsonContent.main === "string") {
```
`else if` ? do you want it to fall back to js and result in an error of unsupported file extension or just can not resolve module? i think we should not pick the js file if there is a .ts one.

billti replied:
Not sure I get your comment or the `else if` prefix. Are you saying only try to load the file referenced by `main` if there was no `typings` setting?

I do notice one change I should make on review though: I should still use the `supportedExtensions` on the base path (i.e. if `main` points to `file1.js`, still look for `file1.d.ts` and `file1.ts` first (and also don't look for and load `.js` if not in the supportedExtensions list)). I'll update for that. Please clarify if I missed your original point.

billti commented:
Simplified the code by using an existing flag, and corrected a couple of minor nits. Also added the logic to not compile/emit any source code that was found by searching the `node_modules` hierarchy.

I'll play with this a little more and add some further tests, but would appreciate a glance over to see if I've overlooked some behavior or requirement somewhere.

billti commented:
Closing this, as it's not going to make 1.8. I have a pull request for `master` I'll open instead.
